napari/_vispy/layers/vectors.py

```diff
@@ -107,25 +110,127 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
     triangles : (2N, 3) array
         Vertex indices that form the mesh triangles
     """
-    ndim = vectors.shape[2]
-    vectors = np.reshape(copy(vectors), (-1, ndim))
-    vectors[1::2] = vectors[::2] + length * vectors[1::2]
 
-    centers = np.repeat(vectors, 2, axis=0)
-    offsets = segment_normal(vectors[::2, :], vectors[1::2, :], p=p)
+    nvectors = vectors.shape[0]
+
+    if style == 'rectangle':
+        vertices = generate_meshes_rectangle_2D(vectors, width, length, p)
+        triangles = np.array(
+            [
+                [2 * i, 2 * i + 1, 2 * i + 2]
+                if i % 2 == 0
+                else [2 * i - 1, 2 * i, 2 * i + 1]
+                for i in range(2 * nvectors)
+            ]
+        ).astype(np.uint32)
+
+    elif style == 'triangle':
+        vertices = generate_meshes_triangle_2D(vectors, width, length, p)
+        triangles = np.arange(3 * nvectors).reshape((-1, 3)).astype(np.uint32)
+
+    elif style == 'arrow':
+        vertices = generate_meshes_arrow_2D(vectors, width, length, p)
+        triangles = np.array(
+            [
+                [7 * i / 3, 7 * i / 3 + 1, 7 * i / 3 + 2]
+                if i % 3 == 0
+                else [
+                    7 * (i - 1) / 3 + 1,
+                    7 * (i - 1) / 3 + 2,
+                    7 * (i - 1) / 3 + 3,
+                ]
+                if i % 3 == 1
+                else [
+                    7 * (i - 2) / 3 + 4,
+                    7 * (i - 2) / 3 + 5,
+                    7 * (i - 2) / 3 + 6,
+                ]
+                for i in range(3 * nvectors)
+            ]
+        ).astype(np.uint32)
```

Comment:
black magic xD Is it possible to add some comments so this is easier to maintain for others in the future?

Reply:
Thank you for your positive review. 

- I am not (yet) well versed in all the typing conventions. I'll investigate `Blending` and make the corresponding changes.
- Please feel free to suggest better names for the styles ! "Rectangles" and "triangles" were chosen to be as objective as possible, but they definitely sound a little boring... ;)
- I also think that `style` would be simpler but wanted to stick to what was done in the rest of the code (I took inspiration from `edge_color_mode`). I'll rename to `style`.
- The functions to create the vertices and triangles are definitely under commented, but I wanted to "pre-push" the PR while I was thinking about the docstrings and comments. For the docstrings, I just discovered the one from the class `InteractionBoxHandle` in  `/layers/base/_base_constants.py` which is what I should aim for in terms of clarity of explanation for my choice of vertices.

Reply:
You did a great job at sticking to the style of the file, I just happen to disagree with it already, so that's on me :) And yeah, I always pre-push without comments at first just to make sure I'm on the right track. Now it's time to actually check the math to avoid broken edge cases, so comments are gonna be useful for reviewing.

> For the docstrings, I just discovered the one from the class InteractionBoxHandle in /layers/base/_base_constants.py which is what I should aim for in terms of clarity of explanation for my choice of vertices.

Sounds good!